RFC-0004-sparse-fill-value/README.md

```diff
@@ -0,0 +1,613 @@
+       tensors as elements. The `fill_value` attribute represents an unspecified
+       element of the array, hence, the fill value has all the
+       properties that any array element has, including the shape and
+       memory consumption.
+
+    2. A non-scalar fill value of a hybrid tensor is more memory-efficient
+       than if the fill value were a scalar.
+
+       For instance, when applying a row-wise function, say, `softmax`
+       to a hybrid tensor, each row may induce a different fill value.
+       As an example, consider a 2D hybrid tensor with one sparse and
+       one dense dimension in COO format, and apply softmax along the
+       first dimension assuming the default fill value `0`. We have:
+
+       ```python
+       >>> A = torch.sparse_coo_tensor(indices=[[0, 3]],
+                                       values=[[.11, .12], [.31, .32]],
+                                       size=(4, 2))
+       >>> A.to_dense()
+       tensor([[0.1100, 0.1200],
+               [0.0000, 0.0000],
+               [0.0000, 0.0000],
+               [0.3100, 0.3200]])
+       >>> A.to_dense().softmax(0)
+       tensor([[0.2492, 0.2503],
+               [0.2232, 0.2220],
+               [0.2232, 0.2220],
+               [0.3044, 0.3057]])
+       ```
+
+       Observe that the softmax result can be represented as a sparse
+       tensor with the same sparsity structure (same indices) as the
+       input only when the fill value is a 1-D dense tensor, that is:
+
+       ```python
+       A.softmax(0) == torch.sparse_coo_tensor(indices=A.indices(),
+                                               values=[[0.2492, 0.2503], [0.3044, 0.3057]],
+                                               size=A.size(),
+                                               fill_value=[0.2232, 0.2220])
+       ```
+
+       The alternative, if the fill value was defined as scalar,
+       would be memory inefficient because the result of
+       `A.softmax(0)` is a full sparse tensor with all elements
+       specified.
+
+6.  The fill value specified that is specified in the `fill_value`
+    argument of the sparse tensor constructors may have different
+    (smaller) shape than the fill value of the hybrid tensor (as
+    defined by point 5).
+
+    The specified fill value (after converting it to `torch.Tensor`
+    instance) can be acquired via `_fill_value()` method.
+
+    For example, the fill value of a (1+1)-D hybrid tensor can be
+    specified as scalar:
+
+    ```python
+    A = torch.sparse_coo_tensor(indices=[[0, 3]],
+                                values=[[.11, .12], [.31, .32]],
+                                size=(4, 2),
+                                fill_value = 1.2)
+    A.fill_value() -> torch.tensor([1.2, 1.2])
+    A._fill_value() -> torch.tensor(1.2)
+    ```
+
+    The output of `fill_value()` is computed as
+
+    ```python
+    A._fill_value().resize(A.values().shape[1:])
```

Comment:
This would introduce undefined elements. Also I cannot find a `resize` method in the docs, only [`Tensor.resize_`](https://pytorch.org/docs/stable/tensors.html?highlight=resize#torch.Tensor.resize_), which would do what I said. I may be missing something though.

Reply:
Right, resize will produce garbage. It turns out that `torch.Tensor.resize` exists but is not documented.
The correct method to use here is `expand`. Will fix it..

Reply:
Fixed.